### ISSUE TITLE
Getting "Request contains an invalid argument" on script v1 requests

### ISSUE THREAD

**Issue by p-doyle** [googleapis/google-api-python-client#628]
#### Environment details

  - OS: Windows 10 and Ubuntu 16.04
  - Python version: 2.7.12
  - pip version: 18.1
  - `google-api-python-client` version: 1.7.4 and 1.7.8

I've been using the script API for several months but within the last 2 days I am unable to even build the service as I keep getting "Request contains an invalid argument".
```

store = Storage('oauth_creds.json')
credentials = store.get()
http = credentials.authorize(httplib2.Http())
service = discovery.build('script', 'v1', http=http)
```

Traceback is:
```
Traceback (most recent call last):
  File "D:\myscript.py", line 1000, in get_script_service
    service = discovery.build('script', 'v1', http=http)
  File "C:\Python27\lib\site-packages\googleapiclient\_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "C:\Python27\lib\site-packages\googleapiclient\discovery.py", line 230, in build
    raise e
```

**Comment by greg-mckeon** [comment]
We're hitting this too.  The issue seems to be that when you pass in an authenticated http, the library uses it even though the discovery api now gives a 400 error if you send a bearer token.  We worked around it by changing L212-214 in discovery.py from
`if http is None:
    discovery_http = build_http()
  else:
    discovery_http = http
`
to
`
discovery_http = http
`

**Comment by p-doyle** [comment]
I tried that but it didn't work for me.  Not sure why I had it like this originally but I ended up changing 
```
http = credentials.authorize(httplib2.Http())
service = discovery.build('script', 'v1', http=http)
```
to this:
`service = discovery.build('script', 'v1', credentials=credentials)`

and it's working fine now.

**Comment by thehesiod** [comment]
yep, I just had to push a fix for this, replacing `discovery_http` with `build_http()` here:

https://github.com/googleapis/google-api-python-client/blob/master/googleapiclient/discovery.py#L223

**Comment by majestique** [comment]
As noted in the referenced issue, this may not be a long term solution since oauth2client is deprecated: https://google-auth.readthedocs.io/en/latest/oauth2client-deprecation.html

**Comment by thehesiod** [comment]
Ya, and unfortunately gauth doesn't support token storage

**Comment by yahanvesh** [comment]
Im getting socket timed out error even after making the  change to 
`
`service = discovery.build('script', 'v1', credentials=credentials)``
**Error**
`  File "/home/admin/automation/googledriveautomation/getediturls.py", line 158, in <module>
    main()
  File "/home/admin/automation/googledriveautomation/getediturls.py", line 103, in main
    service = discovery.build('script', 'v1', credentials=creds)
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/googleapiclient/discovery.py", line 224, in build
    requested_url, build_http(), cache_discovery, cache)
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/googleapiclient/discovery.py", line 271, in _retrieve_discovery_doc
    resp, content = http.request(actual_url)
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/httplib2/__init__.py", line 1514, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "/home/miadmin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/httplib2/__init__.py", line 1264, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/httplib2/__init__.py", line 1187, in _conn_request
    conn.connect()
  File "/home/admin/.pyenv/versions/3.6.4/lib/python3.6/site-packages/httplib2/__init__.py", line 1011, in connect
    sock.connect((self.host, self.port))
socket.timeout: timed out`

Anyone else facing this issue?

**Comment by sripakdp** [comment]
@yahanvesh 

Recently I am getting

 ssl.SSLError: ('The read operation timed out',)

This happened after changing the code to

service = discovery.build('script', 'v1', credentials=credentials)

I think this is an issue on the server side.

**Comment by yahanvesh** [comment]
@sripakdp Im facing the same issue, but unfortunately my scripts take longer times. Anyone who can help get this resolved?

**Comment by yahanvesh** [comment]
@sripakdp  and others who might have faced same issue.

The timeout atleast in my case happened always, even before actual execution of the script took place. Primarily the problem is the httplib bundled with googleapiclient, so i have overriden it with latest httplib2 wherever needed. So i made some changes also in googleapiclient and its now **working**. 

httplib2==0.12.1
google-api-python-client==1.7.8
google-auth==1.6.3
google-auth-httplib2==0.0.3
google-auth-oauthlib==0.3.0


**Changes in googleapiclient/discovery.py**
// Im always overriding discovery_http with httplib2.Http() instead of httplib coming from google 
if http is None:
    discovery_http = build_http()
  else:
    discovery_http = http
  **discovery_http = httplib2.Http()**

**My Code:-**
```
from googleapiclient.discovery import build
from httplib2 import Http
CLIENT_SECRET_FILE = '/home/admin/client_secret.json'

  store = oauth_file.Storage('/home/admin/token.json')
  creds = store.get()
  if not creds or creds.invalid:
      flow = client.flow_from_clientsecrets(CLIENT_SECRET_FILE, SCOPES)
      creds = tools.run_flow(flow, store)
  service = build('script', 'v1', credentials=creds)

  try:
    request = {'function': 'test','parameters': 'one','devMode': True}
    t = time.process_time()

// Again i use HTTP from httplib2
    **http=creds.authorize(Http())**
     response = service.scripts().run(body=request,scriptId=SCRIPT_ID).**execute(http=http)**
    if 'error' in response:
            error = response['error']['details'][0]
            print("Script error message: {0}".format(error['errorMessage']))
    
except errors.HttpError as e:
    print(e.content)
```

**Comment by busunkim96** [comment]
With regards to the original issue, I've contacted the Apps Scripts team to ask them to look into their API configuration. Please continue to pass  the credentials in via `credentials=` on the client.

As @greg-mckeon pointed out, this API returns a 400 when a Authorization header is set. This does not seem to happen with other APIs.
```
$ curl --header "Authorization: Bearer token here"  "https://script.googleapis.com/\$discovery/rest?version=v1"
{
  "error": {
    "code": 400,
    "message": "Request contains an invalid argument.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.DebugInfo",
        "detail": "No scopes defined for OAuth: script.googleapis.com/google.discovery.Discovery.GetDiscoveryRest"
      }
    ]
  }
}
```


For comparison, the Safe Browsing API.
```
curl --header "Authorization: Bearer token here" "https://safebrowsing.googleapis.com/\$discovery/rest?version=v4"
```

Internal issue 142678614

---

For the second issue (timeouts on longer scripts), you can change the timeout value to something higher via `socket.setdefaulttimeout(timeout_in_sec)`. The default is 60 seconds.
https://github.com/googleapis/google-api-python-client/blob/0ba1ca4435aab90bea35a155a2875fa0842b8023/googleapiclient/http.py#L80

https://github.com/googleapis/google-api-python-client/blob/0ba1ca4435aab90bea35a155a2875fa0842b8023/googleapiclient/http.py#L1777-L1792

**Comment by GunK194** [comment]
Hi Team, 
I am getting below error when invoking the dataflow job from cloud function - using googleapiclient for parameters
Traceback (most recent call last): File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker.py", line 383, in run_background_function _function_handler.invoke_user_function(event_object) File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker.py", line 217, in invoke_user_function return call_user_function(request_or_event) File "/env/local/lib/python3.7/site-packages/google/cloud/functions/worker.py", line 214, in call_user_function event_context.Context(**request_or_event.context)) File "/user_code/main.py", line 30, in trigger_dataflow response = request.execute() File "/env/local/lib/python3.7/site-packages/googleapiclient/_helpers.py", line 134, in positional_wrapper return wrapped(*args, **kwargs) File "/env/local/lib/python3.7/site-packages/googleapiclient/http.py", line 907, in execute raise HttpError(resp, content, uri=self.uri) googleapiclient.errors.HttpError: <HttpError 400 when requesting https://dataflow.googleapis.com/v1b3/projects/mygcpdataengineerlab/templates:launch?gcsPath=gs%3A%2F%2Fgcp_dataflow_csv_bq_code%2Ftemplates&alt=json returned "(2452c9f936a8d14a): The workflow could not be created. Causes: (2452c9f936a8d9b7): Found unexpected parameters: ['input' (perhaps you meant 'update'), 'output' (perhaps you meant 'update')]">

code that i used is 
template_body = {
        "jobName": job,
		"parameters": {
            "input": "gs://intient_output/measurementunit.csv",
            "output": "mygcpdataengineerlab:intientpoc.measurementunit",
			
        },
        "environment": {
            "tempLocation": "gs://gcp_dataflow_csv_bq_code/temp",
            }
    }
. Can you guide how to pass it in the main.py file?

**Comment by busunkim96** [comment]
Hi @GunK194. That looks like an unrelated issue to me. Could you open a separate issue?

**Comment by resmungo** [comment]
> I tried that but it didn't work for me. Not sure why I had it like this originally but I ended up changing
> 
> ```
> http = credentials.authorize(httplib2.Http())
> service = discovery.build('script', 'v1', http=http)
> ```
> 
> to this:
> `service = discovery.build('script', 'v1', credentials=credentials)`
> 
> and it's working fine now.

Just wanted to say that I just had this issue and this fixed it. Thank you.

**Comment by busunkim96** [comment]
This will be resolved in #1063

**Comment by parthea** [comment]
I'm going to close this issue since it will be resolved with #1063.
